### PR TITLE
Parse entire HTTP chunk size #396

### PR DESCRIPTION
**Pull request by SkyTrix**
Continue parsing HTTP chunk size up until Int.MaxValue so the actual size of the chunk can be logged if it exceeds the configured max chunk size.

### CODE REVIEW

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala
Comment (pjfanning) [inline http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:323]: Can you include the text `Int.MAX_VALUE` in this message too? Some users may not recognise the numeric value and what it means - still possibly useful to have the actual number too.

Comment (SkyTrix) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:323]: Sure. I believe it should either be `Int.MaxValue` or `Integer.MAX_VALUE` though. Probably the latter due to Java interoperability?

Comment (pjfanning) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:323]: Integer.MAX_VALUE might be better

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala
Comment (He-Pin) [inline http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:323]: I don't think we can have such a big chunk in memory, and if we do want to know how many bytes still there in the chunk, we can just logging:` there is still $n bytes remaining maybe maybe more, increase `settings.maxChunkSize` if you want to parse the full chunk`.

Parse it until the `Int.MaxValue` seems some kind of ddos for me, @jrudolph cc

Comment (SkyTrix) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:323]: Note that this does not start parsing the actual chunk body. It only reads the size value that's written in the chunk header and bails if that's larger than `Int.MaxValue`. It will never attempt to parse anything larger than `settings.maxChunkSize`.

Comment (He-Pin) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:323]: And eg, In Netty, there is a `maxContentLength` to be configured and user can config how to response the when oversized with `handleOversizedMessage`.see : `HttpObjectAggregator`

Comment (He-Pin) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:323]: I did another around of check locally,it's ok.

**Review by mdedetrich — APPROVED** [review]
This seems good to me

**Review by pjfanning — APPROVED** [review]
lgtm

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala
Comment (He-Pin) [inline http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:315]: I see now, It was try decoding how big the chunk size is, and it will stop parsing how if the size already exceed MaxValue.

Comment (He-Pin) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:315]: How about `HTTP chunk size:$size exceeds the configured limit of ${settings.maxChunkSize} bytes`

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala
Comment (He-Pin) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:323]: In netty, when oversize, it invoke a protect method `handleOversizedMessage`, maybe we can introduce something like that too.

**Review by He-Pin — APPROVED** [review]
lgtm after check the code in an IDE

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala
Comment (samueleresca) [inline http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:310]: My personal preference would be to have:
```
if (size > Int.MaxValue) {
// early exit here with failEntityStream
}

// everything else

```

But I don't feel is binding.

**Review by samueleresca — APPROVED** [review]
Looks good to me

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala
Comment (Roiocam) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:310]: guard clauses. 👍🏻

File: http-core/src/test/scala/org/apache/pekko/http/impl/engine/parsing/RequestParserSpec.scala
Comment (jxnu-liguobin) [inline http-core/src/test/scala/org/apache/pekko/http/impl/engine/parsing/RequestParserSpec.scala:548]:
```suggestion
          "400000") should generalMultiParseTo(
```

File: http-core/src/test/scala/org/apache/pekko/http/impl/engine/parsing/RequestParserSpec.scala
Comment (laglangyue) [inline http-core/src/test/scala/org/apache/pekko/http/impl/engine/parsing/RequestParserSpec.scala:537]: "1a2b3c4d5e"

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala
Comment (SkyTrix) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/parsing/HttpMessageParser.scala:310]: I'm trying to keep the structure consistent with the rest of the file. I also think the above suggestion would be incorrect in the sense that `everything else` would still be executed after calling `failEntityStream`? An else block would still be required.

File: http-core/src/test/scala/org/apache/pekko/http/impl/engine/parsing/RequestParserSpec.scala
Comment (SkyTrix) [reply http-core/src/test/scala/org/apache/pekko/http/impl/engine/parsing/RequestParserSpec.scala:548]: I found that the trailing newline is significant in this and many other tests in the file.

**Review by SkyTrix — COMMENTED** [review]
Could a maintainer please shed some light on what it would still take to get this merged? There are currently 6 approvals but also some comments. I don't want to invalidate approvals by making new changes if that wouldn't be required.

**Comment by pjfanning** [comment]
Merged - thanks.

It can be refactored in new PRs.

**Pull request merged**
